worlds/mmx3/Rules.py

```diff
@@ -0,0 +1,451 @@
+                 lambda state: (
+                     state.has(EventName.dr_doppler_lab_1_clear, player) and 
+                     state.has(EventName.dr_doppler_lab_2_clear, player) and 
+                     state.has(EventName.dr_doppler_lab_3_clear, player)
+                    ))
+    else:
+        set_rule(multiworld.get_entrance(f"{RegionName.dr_doppler_lab_1} -> {RegionName.dr_doppler_lab_2}", player),
+                 lambda state: state.has(EventName.dr_doppler_lab_1_clear, player))
+        set_rule(multiworld.get_entrance(f"{RegionName.dr_doppler_lab_2} -> {RegionName.dr_doppler_lab_3}", player),
+                 lambda state: state.has(EventName.dr_doppler_lab_2_clear, player))
+        set_rule(multiworld.get_entrance(f"{RegionName.dr_doppler_lab_3_boss} -> {RegionName.dr_doppler_lab_4}", player),
+                 lambda state: state.has(EventName.dr_doppler_lab_3_clear, player))
+        
+    # Set Boss rematch rules
+    if world.options.doppler_lab_3_boss_rematch_count.value > 0:
+        set_rule(multiworld.get_entrance(f"{RegionName.dr_doppler_lab_3_rematches} -> {RegionName.dr_doppler_lab_3_after_rematches}", player),
+                lambda state: state.has(EventName.boss_rematch_clear, player, world.options.doppler_lab_3_boss_rematch_count.value))
+    
+    # Vile entrance rules
+    vile_open = world.options.vile_open.value
+    entrance_blizzard = multiworld.get_entrance(f"{RegionName.blizzard_buffalo_start} -> {RegionName.vile}", player)
+    entrance_volt = multiworld.get_entrance(f"{RegionName.volt_catfish_start} -> {RegionName.vile}", player)
+    entrance_crush = multiworld.get_entrance(f"{RegionName.crush_crawfish_start} -> {RegionName.vile}", player)
+    if len(vile_open) == 0:
+        set_rule(entrance_blizzard, lambda state: state.has(ItemName.stage_vile, player))
+        set_rule(entrance_volt, lambda state: state.has(ItemName.stage_vile, player))
+        set_rule(entrance_crush, lambda state: state.has(ItemName.stage_vile, player))
+    else:
+        if "Medals" in vile_open and world.options.vile_medal_count.value > 0:
+            add_rule(entrance_blizzard, lambda state: state.has(ItemName.maverick_medal, player, world.options.vile_medal_count.value))
+            add_rule(entrance_volt, lambda state: state.has(ItemName.maverick_medal, player, world.options.vile_medal_count.value))
+            add_rule(entrance_crush, lambda state: state.has(ItemName.maverick_medal, player, world.options.vile_medal_count.value))
+        if "Weapons" in vile_open and world.options.vile_weapon_count.value > 0:
+            add_rule(entrance_blizzard, lambda state: state.has_group("Weapons", player, world.options.vile_weapon_count.value))
+            add_rule(entrance_volt, lambda state: state.has_group("Weapons", player, world.options.vile_weapon_count.value))
+            add_rule(entrance_crush, lambda state: state.has_group("Weapons", player, world.options.vile_weapon_count.value))
```

Comment:
Consider using `has_group_unique` for these, which is like `has_group` but it ignores duplicates.
I have a feeling these will erroneously succeed if someone start_inventory's something then receives a duplicate of it.

Reply:
Ah, I didn't consider it. I just copied this from the Ride Armor logic and didn't think about it twice.